### PR TITLE
Chicago Crimes Example

### PR DESCRIPTION
**Pull request by bodo-sai**

### CODE REVIEW

**Review by alifarhid — CHANGES_REQUESTED** [review]
Please add the followings:
1. instruction on how to run locally on Docker. 
2. Test end to end on AWS EKS. There are problems there that needs to be fixed. Not sure what are the problems.

File: Kubernetes/README.md [hunk 1]
```diff
@@ -8,60 +8,14 @@ Bodo workloads can be deployed with Kubernetes using the [Kubeflow MPI-Operator]
...
-You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [pi calculation example](docker/pi.py), which is used in this tutorial.
+You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [Chicago Crimes example](docker/chicago_crimes.py), which is used in this tutorial.
 In case of private registries, follow instructions from [here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).
 
 
 ## Setup
-Bodo can be deployed in any Kubernetes cluster. For the purposes of this example, we set up a Kubernetes Cluster in EKS using KOPS:
+Bodo can be deployed in any Kubernetes cluster. For the purposes of this example, we will be using a Kubernetes Cluster in EKS:
 
-### Step 1: Create a Kubernetes Cluster in EKS
```
Comment (sahil1105) [inline Kubernetes/README.md:18]: Why remove this step? We could mark it as optional instead? Would be good as a reference for people with limited K8s experience.

File: Kubernetes/README.md [hunk 2]
```diff
@@ -8,60 +8,14 @@ Bodo workloads can be deployed with Kubernetes using the [Kubeflow MPI-Operator]
 
 - Create a docker image that contains your intended Bodo version and Python scripts and upload it to a docker registry such as Docker Hub so that K8s can pull it. For reference, see this [Dockerfile](docker/Dockerfile).
 A docker image created from this Dockerfile is also available on DockerHub: [bodoaidocker/kube-mpioperator-minimal](https://hub.docker.com/r/bodoaidocker/kube-mpioperator-minimal/tags).
-You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [pi calculation example](docker/pi.py), which is used in this tutorial.
+You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [Chicago Crimes example](docker/chicago_crimes.py), which is used in this tutorial.
```
Comment (sahil1105) [inline Kubernetes/README.md:11]: I think we can include both examples. `pi.py` is easier for validation purposes.

File: Kubernetes/README.md [hunk 3]
```diff
@@ -84,15 +38,20 @@ NAME                   CREATED AT
 mpijobs.kubeflow.org   2022-01-03T21:19:10Z
 ```
 
-### Step 3: Run your Bodo application
+### Step 2: Run your Bodo application
+
+- Define a kubernetes resource for your Bodo workload, such as the one defined in [`mpijob.yaml`](mpijob.yaml) that runs the [Chicago Crimes example](docker/chicago_crimes.py). You can modify it based on your cluster configuration: 
 
-- Define a kubernetes resource for your Bodo workload, such as the one defined in [`example-mpijob.yaml`](example-mpijob.yaml) that runs the [pi calculation example](docker/pi.py). You can modify it based on your cluster configuration: update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node and set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster. Lastly, make sure `-n` is equal to `spec.mpiReplicaSpecs.Worker.replicas` multiplied by `spec.slotsPerWorker`, i.e. the total number of physical cores on your worker nodes. If you're using the cluster configuration as defined in step 1, you do not need to modify anything.
+1. update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node
```
Comment (sahil1105) [inline Kubernetes/README.md:45]:
```suggestion
1. Update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node.
```

File: Kubernetes/README.md [hunk 4]
```diff
@@ -84,15 +38,20 @@ NAME                   CREATED AT
...
-### Step 3: Run your Bodo application
+### Step 2: Run your Bodo application
+
+- Define a kubernetes resource for your Bodo workload, such as the one defined in [`mpijob.yaml`](mpijob.yaml) that runs the [Chicago Crimes example](docker/chicago_crimes.py). You can modify it based on your cluster configuration: 
 
-- Define a kubernetes resource for your Bodo workload, such as the one defined in [`example-mpijob.yaml`](example-mpijob.yaml) that runs the [pi calculation example](docker/pi.py). You can modify it based on your cluster configuration: update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node and set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster. Lastly, make sure `-n` is equal to `spec.mpiReplicaSpecs.Worker.replicas` multiplied by `spec.slotsPerWorker`, i.e. the total number of physical cores on your worker nodes. If you're using the cluster configuration as defined in step 1, you do not need to modify anything.
+1. update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node 
+2. set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster.
```
Comment (sahil1105) [inline Kubernetes/README.md:46]:
```suggestion
2. Set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster. 
```

File: Kubernetes/README.md [hunk 5]
```diff
@@ -84,15 +38,20 @@ NAME                   CREATED AT
...
 
-- Define a kubernetes resource for your Bodo workload, such as the one defined in [`example-mpijob.yaml`](example-mpijob.yaml) that runs the [pi calculation example](docker/pi.py). You can modify it based on your cluster configuration: update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node and set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster. Lastly, make sure `-n` is equal to `spec.mpiReplicaSpecs.Worker.replicas` multiplied by `spec.slotsPerWorker`, i.e. the total number of physical cores on your worker nodes. If you're using the cluster configuration as defined in step 1, you do not need to modify anything.
+1. update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node 
+2. set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster. 
+3. Build the image using the Dockerfile and replace the image at `spec.mpiReplicaSpecs.Launcher.template.spec.containers.image` and  `spec.mpiReplicaSpecs.Worker.template.spec.containers.image`.
```
Comment (sahil1105) [inline Kubernetes/README.md:47]: I would also upload an image to our DockerHub, so people can test more easily.

File: Kubernetes/docker/pi.py [hunk 6]
```diff
@@ -1,33 +0,0 @@
-"""
```
Comment (sahil1105) [inline Kubernetes/docker/pi.py:1]: I'd keep this as well.

File: Kubernetes/docker/Dockerfile [hunk 7]
```diff
@@ -2,16 +2,20 @@ FROM mpioperator/intel:latest
...
-# Install Miniconda
 RUN apt-get update && apt-get install -y wget bzip2 git vim make curl\
-	&& wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh \
-	&& chmod +x miniconda.sh\
-	&& ./miniconda.sh -p /home/mpiuser/miniconda3 -b
-ENV PATH /home/mpiuser/miniconda3/bin/:${PATH}
+        && wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh \
+        && chmod +x miniconda.sh\
+        && ./miniconda.sh -b
+ENV PATH /root/miniconda3/bin/:${PATH}
```
Comment (sahil1105) [inline Kubernetes/docker/Dockerfile:9]: Are we sure `root` works in all cases? MPI usually doesn't like to be run using the root user.

File: Kubernetes/docker/chicago_crimes.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    print(get_crimes_type_dates.head())
+
+print("A general view of crime records by time, type and location")
+print("Determining the pattern on daily basis")
+
+crimes_days = get_crimes_by_days(crimes)
+if bodo.get_rank()==0:
+    print(crimes_days.head())
+
+print("Determining the pattern on monthly basis\n")
```
Comment (sahil1105) [inline Kubernetes/docker/chicago_crimes.py:127]: Add this and everything below in a `main` function and then call using `if __name__ == '__main__'`?

File: Kubernetes/docker/chicago_crimes.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+@bodo.jit(distributed=['crimes', 'crimes_location'], cache=True)
+def get_crimes_by_location(crimes):
+    t1 = time.time()
+    crimes_location = crimes.groupby('Location Description', as_index=False)['ID'].count().sort_values(by='ID', ascending=False)
+    print("Group by days Time: ", ((time.time() - t1) * 1000), " (ms)")
+    return crimes_location
+
+
+print(f"Hello World from rank {bodo.get_rank()}. Total ranks={bodo.get_size()}")
```
Comment (sahil1105) [inline Kubernetes/docker/chicago_crimes.py:90]: Add this and everything below in a `main` function and then call using `if __name__ == '__main__'`?

File: Kubernetes/mpijob.yaml [hunk 10]
```diff
@@ -1,7 +1,7 @@
-apiVersion: kubeflow.org/v2beta1
+apiVersion: kubeflow.org/v1
```
Comment (sahil1105) [inline Kubernetes/mpijob.yaml:1]: Revert?

File: Kubernetes/mpijob.yaml [hunk 11]
```diff
@@ -14,7 +14,7 @@ spec:
       template:
         spec:
           containers:
-            - image: bodoaidocker/kube-mpioperator-minimal:latest
+            - image: <YOUR-DOCKER-IMAGE-HERE>
```
Comment (sahil1105) [inline Kubernetes/mpijob.yaml:17]: Same as above, I would upload one to our DockerHub to make testing easier.

**Review by sahil1105 — COMMENTED** [review]
Thanks @bodo-sai!
As discussed, we also need to get a better understanding of how we resource allocation can be managed and ensured, and then document it.

File: Kubernetes/README.md (hunk 1 above)
Comment (bodo-sai) [reply Kubernetes/README.md:18]: Yes, we can it as an Optional one at the end maybe.

File: Kubernetes/README.md (hunk 2 above)
Comment (bodo-sai) [reply Kubernetes/README.md:11]: But @alifarhid pointed out that pi.py is not a good explanation of what bodo is capable of

File: Kubernetes/docker/Dockerfile (hunk 7 above)
Comment (bodo-sai) [reply Kubernetes/docker/Dockerfile:9]: We are not using root for execution, we are just installing using root user and providing access to root folder for mpi user

File: Kubernetes/mpijob.yaml (hunk 10 above)
Comment (bodo-sai) [reply Kubernetes/mpijob.yaml:1]: It's a mistake from my end

**Review by alifarhid — APPROVED** [review]

File: Kubernetes/docker/Dockerfile [hunk 12]
```diff
@@ -2,16 +2,20 @@ FROM mpioperator/intel:latest
...
 
-RUN conda install -y python=3.9 bodo=2021.11 -c bodo.ai -c conda-forge
+RUN conda install -y mamba -c conda-forge && mamba install -y bodo -c bodo.ai -c conda-forge
```
Comment (sahil1105) [inline Kubernetes/docker/Dockerfile:11]: I would fix the bodo version for reproducibility purposes. Or add a note that this was tested with so and so version.

File: Kubernetes/README.md [hunk 13]
```diff
@@ -6,9 +6,14 @@ Bodo workloads can be deployed with Kubernetes using the [Kubeflow MPI-Operator]
...
+- Create a docker image that contains your intended Bodo version and Python scripts and upload it to a docker registry such as Docker Hub so that K8s can pull it. 
+For reference, see this [Dockerfile](docker/Dockerfile). We have provided two python scripts which support bodo.
+
+1. `pi.py` which is used for basic testing or validation purposes
+2. `chicago_crimes.py` whic can used to test bodo on a larger dataset. 
+
 A docker image created from this Dockerfile is also available on DockerHub: [bodoaidocker/kube-mpioperator-minimal](https://hub.docker.com/r/bodoaidocker/kube-mpioperator-minimal/tags).
-You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [Chicago Crimes example](docker/chicago_crimes.py), which is used in this tutorial.
+You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [pi calculation example](docker/chicago_crimes.py), which is used in this tutorial.
```
[the remark this answers is not in the document]
Comment (alifarhid) [reply Kubernetes/README.md:16]: If it is supposed to be pi calculation, please also fix the docker/chicago_crimes.py to docker/pi.py. Not sure. @bodo-sai please check.

File: Kubernetes/docker/Dockerfile (hunk 12 above)
Comment (alifarhid) [reply Kubernetes/docker/Dockerfile:11]: @bodo-sai please check. I already saw you used mamba in the new docker file. Not sure why it's back to older one.

File: Kubernetes/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (bodo-sai) [reply Kubernetes/README.md:16]: It's a type

**Pull request merged**